lgc/builder/BuilderImpl.cpp

```diff
@@ -172,13 +181,53 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
           indices = ArrayRef<int>({2, 3});
           input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), indices);
           input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), indices);
-          product =
+          scalar =
               CreateIntrinsic(intrinsicDot2, {}, {input1, input2, intermediateRes, getInt1(false)}, nullptr, instName);
         }
         // Add a satruation add if required.
-        if (!hasNoAccumulator)
-          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {product, accumulator}, {}, instName);
+        if (hasAccumulator)
+          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {scalar, accumulator}, {}, instName);
+      }
+    }
+  }
+
+  // Do clamp if it has an accumulator
+  // NOTE: 32-bit result is a saturating add result. Do the manual saturating for 8-bit/16-bit result.
+  if (scalar->getType() != outputTy) {
+    if (hasAccumulator) {
+      // The scope for 8-bit/32-bit/16-bit signed/unsinged integer type
+      struct ClampRange {
+        long long signedMin;
+        long long signedMax;
+        unsigned long long unsignedMin;
+        unsigned long long unsignedMax;
+      };
+
+      ClampRange range = {};
+      const unsigned bitWidth = outputTy->getScalarSizeInBits();
```

Comment:
Thanks for changing to calculate these. But the source could be shortened even more by not having "struct ClampRange" and just having four auto variables.